plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/codemodernizer/client/GumbyClient.kt

```diff
@@ -101,48 +109,60 @@ class GumbyClient(private val project: Project) {
-                codeTransformApiNames = apiName,
-                codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
-                codeTransformRunTimeLatency = calculateTotalLatency(startTime, Instant.now()),
+            telemetry.logApiLatency(
+                apiName,
+                startTime,
                 codeTransformUploadId = uploadId,
                 codeTransformJobId = jobId,
-                codeTransformRequestId = result?.responseMetadata()?.requestId(),
+                codeTransformRequestId = result?.responseMetadata()?.requestId()
             )
         }
     }
 
     suspend fun downloadExportResultArchive(jobId: JobId): MutableList<ByteArray> = amazonQStreamingClient.exportResultArchive(
         jobId.id,
         ExportIntent.TRANSFORMATION,
-        {
-                e ->
+        { e ->
             LOG.error(e) { "${CodeTransformApiNames.ExportResultArchive} failed: ${e.message}" }
-            CodetransformTelemetry.logApiError(
-                codeTransformApiNames = CodeTransformApiNames.ExportResultArchive,
-                codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
-                codeTransformApiErrorMessage = e.message.toString(),
-                codeTransformJobId = jobId.id,
-            )
+            telemetry.apiError(e.localizedMessage, CodeTransformApiNames.ExportResultArchive, jobId.id)
         },
         { startTime ->
-            CodetransformTelemetry.logApiLatency(
-                codeTransformApiNames = CodeTransformApiNames.ExportResultArchive,
-                codeTransformSessionId = CodeTransformTelemetryState.instance.getSessionId(),
-                codeTransformRunTimeLatency = calculateTotalLatency(startTime, Instant.now()),
-                codeTransformJobId = jobId.id,
-            )
+            telemetry.logApiLatency(CodeTransformApiNames.ExportResultArchive, startTime, codeTransformJobId = jobId.id)
         }
     )
 
+    /*
+     * Adapted from [CodeWhispererCodeScanSession]
+     */
+    fun uploadArtifactToS3(url: String, fileToUpload: File, checksum: String, kmsArn: String, shouldStop: () -> Boolean) {
+        HttpRequests.put(url, APPLICATION_ZIP).userAgent(AwsClientManager.userAgent).tuner {
+            it.setRequestProperty(CONTENT_SHA256, checksum)
+            if (kmsArn.isNotEmpty()) {
+                it.setRequestProperty(SERVER_SIDE_ENCRYPTION, AWS_KMS)
+                it.setRequestProperty(SERVER_SIDE_ENCRYPTION_AWS_KMS_KEY_ID, kmsArn)
+            }
+        }
+            .connect { request -> // default connect timeout is 10s
+                val connection = request.connection as HttpURLConnection
+                connection.setFixedLengthStreamingMode(fileToUpload.length())
+                fileToUpload.inputStream().use { inputStream ->
+                    connection.outputStream.use {
+                        val bufferSize = 4096
```

Comment:
Did we change anything about this upload function & also is 4096 big enough?

Reply:
Not in this PR, just moved it. 
For chunk size, we could increase it I guess, but honestly we should reconsider the entire upload mechanic and move to multipart uploads.